fastlane/Fastfile

```diff
@@ -605,16 +605,153 @@ platform :ios do
+        derived_data_path: "scan_derived_data",
+        output_types: 'junit',
+        output_style: 'raw',
+        result_bundle: true,
+        testplan: options[:test_plan],
+        configuration: 'Debug',
+        output_directory: "fastlane/test_output/xctest/ios",
+
+        fail_build: false # TODO: Need this for retry but maybe find a way to fail if we don't want this
+      )
+    rescue => ex
+        UI.error("Tests failed: #{ex}")
+    end
+
+    report_path = File.absolute_path('./test_output/xctest/ios/report.junit')
+    FileUtils.cp(report_path, "#{report_path}.original")
+    save_failed_tests(path: report_path)
+  end
+
+  private_lane :failed_tests_path do
+    File.absolute_path('./test_output/xctest/ios/failed_tests.txt')
+  end
+
+  private_lane :save_failed_tests do |options|
+    require 'nokogiri'
+    
+    failed_tests = []
+    report_path = options[:path]
+    
+    if File.exist?(report_path)
+      doc = Nokogiri::XML(File.open(report_path))
+  
+      doc.xpath('//testcase[failure]').each do |test_case|
+        suitename = test_case.parent['name'] # Retrieve the suitename
+        classname = test_case['classname']
+        name = test_case['name']
+        failed_tests << "#{suitename}/#{classname}/#{name}"
+      end
+
+      failed_tests = failed_tests.uniq
+  
+      File.open(failed_tests_path, 'w') do |file|
+        file.puts(failed_tests)
+      end
+    end
+  end
+
+  lane :retry_failed_tests do |options|
+    fail_build = options[:fail_build]
+    retry_attempt = options[:retry_attempt]
+    failed_tests = File.read(failed_tests_path).split("\n")
+    
+    if failed_tests.empty?
+      UI.message('No failed tests to retry')
+    else
+      artifact_path = "./test_output/xctest/ios"
+      report_dir = "test_output_retry_#{retry_attempt}/xctest/ios"
+      report_path = File.absolute_path(File.join(".", report_dir, "report.junit"))
+
+      scan(
+        scheme: "BackendIntegrationTests",
+        derived_data_path: "scan_derived_data",
+        output_types: 'junit',
+        output_style: 'raw',
+        result_bundle: true,
+        configuration: 'Debug',
+        output_directory: File.join('fastlane', report_dir),
+
+        only_testing: failed_tests,
+        fail_build: !!fail_build
+      )
+
+      moved_report_path = File.join(artifact_path, "report.junit.retry.#{retry_attempt}")
+      FileUtils.mv(report_path, moved_report_path)
+
+      save_failed_tests(path: moved_report_path)
+      merge_and_replace_junit(retry_path: moved_report_path)
+    end
+  end
+
+  lane :merge_and_replace_junit do |options|
+    retry_report = options[:retry_path]
+
+    original_report = File.absolute_path('./test_output/xctest/ios/report.junit')
+
+    merged_report = merge_reports(original_report, retry_report)
+    save_report(merged_report, original_report)
+
+    FileUtils.rm_rf('./test_output/xctest/ios/retry')
+
+    puts "Merged report saved to #{original_report}"
```

Comment:
Should this use fastlane log tools like `UI.message`?

Reply:
Ooooops! Yes yes yes 👍